Handle Ensembl transcript versions

davmlaw opened the pull request:
Fix for issue #40 seqfetcher doesn't support Ensembl transcript versions

Feel free to change exception type / message / whatever

davmlaw committed:
Handle Ensembl transcript versions

reece commented on src/bioutils/seqfetcher.py:
```diff
@@ -139,11 +139,26 @@ def _fetch_seq_ensembl(ac, start_i=None, end_i=None):
+    # Ensembl API only takes transcript IDs (without version) and returns the latest one
+    # So we need to strip the transcript version, then check if what was returned was the one we requested
+    version = 0
+    m = re.match(r"^(ENST\d+)\.(\d)$", ac)
```
Versions > 9 exist, right? ie, `\d+` instead.

davmlaw replied:
Yes good catch

reece commented on src/bioutils/seqfetcher.py:
```diff
@@ -139,11 +139,26 @@ def _fetch_seq_ensembl(ac, start_i=None, end_i=None):
+        ac, version = m.groups()
+        version = int(version)
+
     url_fmt = "http://rest.ensembl.org/sequence/id/{ac}"
```
Preferred (by me): `url = f"http://rest.ensembl.org/sequence/id/{ac}"`

davmlaw replied:
So do I - but that's a 3.8 feature - and setup has ```python_requires = >= 3.6``` so that will break 3.6

davmlaw replied:
Ah, looked and it's a 3.6 feature so that's ok

reece reviewed (commented):
Minor comments/questions. Otherwise, looks good to me.
Please reply and I'll merge.
Thanks for the contribution.

davmlaw committed:
Comments from pull request.

Also deal with case of someone passing in version 0 (which is invalid but should be caught as error)

davmlaw commented:
Hi, I committed your changes into my branch/pull request

reece commented on src/bioutils/seqfetcher.py:
```diff
@@ -139,11 +139,25 @@ def _fetch_seq_ensembl(ac, start_i=None, end_i=None):
+        latest_version = data["version"]
+        if version != latest_version:
+            msg = f"Ensembl API only provides {ac} version ({latest_version}), requested: {version}"
+            raise requests.RequestException(msg)
```
@davmlaw : Can we please raise KeyError here instead? The error doesn't have anything to do with http or requests, so an application exception (rather than a protocol exception) seems more appropriate to me.

Sorry I didn't catch this earlier.

davmlaw committed:
Throw KeyError on wrong Ensembl transcript version

davmlaw commented:
Hi, changed to KeyError, and added to list of throws in docstring.

I used Requests as that was the only exception the function threw before, I haven't checked calling code (I suspect that is in other repos etc)

reece commented:
Great. Thanks @davmlaw . Merging.
